### DISCUSSION TITLE
Support for .kml files

### DISCUSSION THREAD

**Discussion by hugo-mcfrojd** [Freika/dawarich#350] [Ideas]
Are there any plans to include support for .kml files?

I have a gps in my RV connected to my Victron electrical system.
From there VRM (online portal) i can choose a date range and then download my tracks in .kml format only.

Today i use these files to create maps on Google My Maps.
But would love to be able to add them to Dawarich.

I have tried the online "kml to gpx" converters but the files outputted there wount import into Dawarich.

[gps_example.zip](https://github.com/user-attachments/files/17475318/gps_example.zip)

**Comment by Freika** [comment]
I'm not sure, is KML a widely known format? Does it have timestamps, or it's made for _planning_ routes only? Need more knowledge here :)

UPD: Opened the attached file, the only timestamp is in the name: `<name>2024-07-11 06:00:35 - 2024-07-12 21:59:38</name>`

I'm not sure about how to map timestamps to the points that are contained in the file, should they be spread evenly?..

**Reply by dunc1ca** [reply]
Kml is a widely used format and I have also had the same issue as Hugo.  I have included a kml file that I downloaded from FlightRadar24 of my flight back in July.  You will see that every point and line has a time stamp associated with it: 
`<TimeStamp>
          <when>2024-07-07T08:41:31+00:00</when>
        </TimeStamp>`

[LO464-36073ee6.zip](https://github.com/user-attachments/files/17478885/LO464-36073ee6.zip)

**Reply by Freika** [reply]
@dunc1ca aha, this one is way more interesting! I'll think about implementing support for this kind of files then, thank you!

**Reply by andreas-kupries** [reply]
KML/KMZ is also the format natively exported by Google Maps.

**Comment by rtuszik** [comment]
+1 on .kml support. I import flights and these usually come in .kml from my source.

**Reply by adromir** [reply]
If I may ask, what do you use to get this data?

**Reply by harrymrh** [reply]
Google Tracks.  A now retired tracking app.

**Comment by adromir** [comment]
@Freika maybe GPSBabbel is worth while to look into, to integrate into dawarich to support prretty much every Fileformat that is available

https://www.gpsbabel.org/capabilities.html

**Reply by Freika** [reply]
I will certainly have a look!

**Reply by Freika** [reply]
@adromir I briefly look at it and it seems like
1. It doesn't support Google's Records.json yet (only semantic history files)
2. And for those, when converting to geojson, it simply removes timestamp attributes, which is unfortunately a no-go for Dawarich.

But thanks for suggestion, it's a nice piece of software to know about!

**Reply by adromir** [reply]
Well my Idea behind it was not to replace the import functionality , but maybe as a "quick and dirty" way to support other filetypes. Like if not GPX or Google History or Records.json then convert to a file format that is already implemented and use that to import.

A quick test with kml -> gpx seem to have worked properly, but i am not an expert on gps files

**Reply by adromir** [reply]
I did some more investigation on this topic and it actually looks like that timestamps are not part of the official GeoJson standard. So it might actually a bit risky to rely on it as a format for imports. Unless you can make sure it's well formed in the specifications you need for the import

**Reply by Freika** [reply]
You might be right, I'll think about it

**Reply by harrymrh** [reply]
I have tried converting my old google tracks .kml files to gpx with gpsbabel.  No points are found.  there is a `<wpt></wpt> `before the points in the `<trk></trk>` section.  Is this the problem? 
`FOR %%i IN (*.kml) DO "C:\Program Files\GPSBabel\gpsbabel.exe" -t -i kml,lines=0,points=0,trackdata=0,labels=0 -f  %%i -x nuketypes,routes,waypoints -x transform,trk=wpt,del -o gpx -F %%~Ni_kml.gpx`
seems to get rid, but I am still at 0 detected points, with many in the gpx file.
What is the specification of a gpx file that makes dawarich happy?

**Reply by adromir** [reply]
Have you tried to convert a single file into a gpx with the GUI of GPSBabel already? Is it also broken? I think the culprit might be setting points=0 but i am not sure about it

**Reply by harrymrh** [reply]
I seem to be getting Import "GoogleTracksTrackpoints.gpx" failed: undefined method 'timestamp' for nil, stacktrace:
not worked out the meaning yet.

**Reply by adromir** [reply]
Sounds like the new data is missing timestamps. Probably trackdata=0 is the culprit. Just try a single conversion without any additional options. Just in and out file

**Reply by harrymrh** [reply]
the kml files import to more than one `<trk> </trk>` section per file, named and with its own 
`    <trkseg>
      <trkpt lat="79.470825000" lon="17.907613000">
        <ele>1997.000</ele>
        <time>2013-11-20T07:30:41.200Z</time>
      </trkpt>`
full selection of points per section, as above, timestamps for everything.  These multiple tracks per file seems to be the difference between a gpx file that works and one that doesn't.  The way the KML files were recorded seems to dictate this structure.  GPX files created from other sources don't seem to have this trouble.

**Reply by adromir** [reply]
What is the time format in the gpx files? Maybe they are not converted into timestamps

**Reply by harrymrh** [reply]
Time format in above example

**Comment by gchai** [comment]
+1 for KML files, I was just on a flight where the recording stopped halfway, probably because of having a bad signal since I was in a metal tube.

Was able to download the KML file for the flight following [these directions](https://www.reddit.com/r/FogofWorld/comments/1483ec8/how_do_i_download_a_flight_path_so_i_can_import_it/jnyq22l/) and I was able to import it into Fog of World.

**Reply by gchai** [reply]
If I wanted to try to contribute, where should I take a look to get started? Specifically the import parsing
